contracts/Remittance.sol

```diff
@@ -1,16 +1,138 @@
+    // Hash is a sha256 hash of the broker address + recipient password
+    function create(bytes32 recipientPasswordHash, address broker, uint256 expiryDate)
+    external
+    payable
+    whenNotPaused
+    returns(bool success)
+    {
+        require(broker != msg.sender, "The caller cannot be the broker");
+        require(msg.value > 0, "The amount must be greater than 0");
+        require(broker != address(0), "Address cannot be zero");
+        checkEmptyHash(recipientPasswordHash);
+
+        RemittanceInstance storage remittanceInstance = remittances[recipientPasswordHash];
+
+        require(remittanceInstance.funder == address(0), "Remittance already exists");
+
+        remittanceInstance.funder = msg.sender;
+        remittanceInstance.broker = broker;
+        remittanceInstance.fundsOwed = msg.value;
+        remittanceInstance.expiryDate = expiryDate;
+        remittanceInstance.valid = true;
+
+        emit RemittanceFundsCreated(recipientPasswordHash, msg.sender, broker, msg.value, expiryDate);
+
+        return true;
+    }
+
+    // The broker sends the recipient's password to release the funds
+    function release(bytes32 recipientPasswordHash) external whenNotPaused returns(bool success) {
+
+        checkEmptyHash(recipientPasswordHash);
+        bytes32 rehash = keccak256(abi.encodePacked(recipientPasswordHash, address(msg.sender)));
+
+        //Retrieve remittance
+        RemittanceInstance storage remittanceInstance = remittances[rehash];
+
+        uint256 fundsOwed = remittanceInstance.fundsOwed;
+
+        require(remittanceInstance.valid, "No remittance found");
+        require(fundsOwed > 0, "No funds available");
+
+        require(block.timestamp < remittanceInstance.expiryDate, "The remittance has expired");
+
+        // Indicate remittance has been collected
+        remittanceInstance.fundsOwed = 0;
+        remittanceInstance.expiryDate = 0;
+
+        emit RemittanceFundsReleased(rehash, msg.sender, fundsOwed);
+
+        (success, ) = msg.sender.call{value: fundsOwed}("");
+        require(success, "Transfer failed");
+    }
+
+    // Funder can retrieve funds if not claimed by the expiry date
+    function reclaim(bytes32 recipientPasswordHash) external whenNotPaused returns(bool success) {
+
+        checkEmptyHash(recipientPasswordHash);
+
+        //Retrieve remittance
+        RemittanceInstance storage remittanceInstance = remittances[recipientPasswordHash];
+
+        uint256 fundsOwed = remittanceInstance.fundsOwed;
+
+        require(remittanceInstance.valid, "No remittance found");
+        require(fundsOwed > 0, "No funds available");
+
+        require(block.timestamp >= remittanceInstance.expiryDate, "The remittance has not expired");
+        require(msg.sender == remittanceInstance.funder, "Only the funder can reclaim funds");
+
+        // Indicate that it was the funder who retrieved the expired funds
+        remittanceInstance.broker = msg.sender;
```

Comment:
Mmh not sure about this one. It's like leaving duplicate info with the event.

Reply:
I will remove it. I was still crystallising my understanding of storage vs events. I have been doing some more reading around. I think I understand that the storage is way to maintain state during the exchange and once the exchange is complete, the storage can then be discarded, reused or zeroed because the transaction history is stored in the logs:

https://ethereum.stackexchange.com/a/6526/48189

Reply:
At the core, a smart contract is only here to check and approve what others do. The storage is here to make information from 1 transaction accessible to another one. To inform the outside world, an event is enough.
Now, if you wanted to implement a reputation system where you want to punish brokers you "fail", perhaps keeping this information is perhaps useless. But perhaps not, because your reputation could very well be updated as it happens.

Reply:
Thank you. I understand much better.